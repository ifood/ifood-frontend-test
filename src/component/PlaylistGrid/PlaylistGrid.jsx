import React from 'react';
import PropTypes from 'prop-types';
import { useIntl } from 'react-intl';
import Playlist from '../Playlist';
import Info from '../Info';
import Grid from '../Grid';

function itemRenderer(playlist) {
  return <Playlist key={playlist?.id} playlist={playlist} />;
}

function PlaylistGrid({ title, items }) {
  const intl = useIntl();

  const gridTitle =
    title != null ? title : intl.formatMessage({ id: 'featured.playlists' });

  if (items == null || items.length === 0) {
    return (
      <Info
        icon="search"
        title={intl.formatMessage({ id: 'no.playlist' })}
        info={intl.formatMessage({ id: 'choose.filters.message' })}
      />
    );
  }

  return <Grid title={gridTitle} items={items} itemRenderer={itemRenderer} />;
}

PlaylistGrid.propTypes = {
  title: PropTypes.string,
  items: PropTypes.array,
};

export default React.memo(PlaylistGrid);
