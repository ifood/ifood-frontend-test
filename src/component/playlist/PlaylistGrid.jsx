/* eslint-disable react-hooks/exhaustive-deps, camelcase */
import React, { useCallback, useEffect, useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {
  Accordion,
  IconButton,
  TextField,
  Typography,
} from '@material-ui/core';
import { FilterList } from '@material-ui/icons';
import GridList from '../gridlist/GridList';
import PlaylistCard from './PlaylistCard';
import { CenteredFlex, LayoutContent } from '../containers/Containers';
import FilterForm from '../filter/FilterForm';
import PlaylistError from './PlaylistError';

const Content = styled(LayoutContent)`
  margin-top: 150px;
`;
const FilterContainer = styled.div`
  display: flex;
  padding-bottom: 16px;
`;
const AccordionContainer = styled.div`
  padding-bottom: 16px;
`;

function itemRenderer(element) {
  return <PlaylistCard playlist={element} />;
}

function PlaylistGrid(props) {
  const {
    data,
    errorMessage,
    filterData,
    filters,
    onChangeFilter,
    title,
  } = props;
  const [filterText, setFilterText] = useState('');
  const [gridData, setGridData] = useState([]);
  const [expandedFilter, setExpandedFilter] = useState(false);

  useEffect(() => {
    setGridData([...data]);
  }, [data]);

  const handleChangeTextFilter = useCallback((element) => {
    setFilterText(element?.target?.value);
  }, []);

  const handleExpandFilter = useCallback(() => {
    setExpandedFilter(!expandedFilter);
  }, [expandedFilter]);

  useEffect(() => {
    const filterValues = data.filter(
      (element) =>
        element.name.toUpperCase().includes(filterText.toUpperCase()) ||
        element.description.toUpperCase().includes(filterText.toUpperCase()),
    );

    setGridData(filterValues);
  }, [filterText]);

  const renderGrid = useMemo(() => {
    if (errorMessage != null) {
      return <PlaylistError error={errorMessage} />;
    }

    if (gridData.length > 0) {
      return (
        <>
          <Typography color="textPrimary" variant="h5" gutterBottom>
            {title}
          </Typography>
          <GridList
            data={gridData}
            itemRenderer={itemRenderer}
            filters={filters}
          />
        </>
      );
    }

    return (
      <CenteredFlex>
        <Typography color="textPrimary" variant="h5" gutterBottom>
          Your search result did not return any playlist
        </Typography>
      </CenteredFlex>
    );
  }, [title, gridData, itemRenderer, filters, errorMessage]);

  return (
    <Content>
      <AccordionContainer>
        <Accordion expanded={expandedFilter}>
          <FilterContainer>
            <TextField
              id="local-filter"
              label="Search playlist"
              variant="filled"
              onChange={handleChangeTextFilter}
              fullWidth
            />
            <IconButton color="primary" onClick={handleExpandFilter}>
              <FilterList />
            </IconButton>
          </FilterContainer>
          <FilterForm
            filters={filters}
            onChange={onChangeFilter}
            filterData={filterData}
          />
        </Accordion>
      </AccordionContainer>
      {renderGrid}
    </Content>
  );
}

PlaylistGrid.defaultProps = {
  data: [],
};

PlaylistGrid.propTypes = {
  data: PropTypes.array,
  title: PropTypes.string,
  errorMessage: PropTypes.string,
  filters: PropTypes.array,
  filterData: PropTypes.object,
  onChangeFilter: PropTypes.func,
};

export default PlaylistGrid;
