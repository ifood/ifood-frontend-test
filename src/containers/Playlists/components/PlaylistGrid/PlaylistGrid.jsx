
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import GridList from '@material-ui/core/GridList';
import GridListTile from '@material-ui/core/GridListTile';
import GridListTileBar from '@material-ui/core/GridListTileBar';

const Container = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
  margin-top: 25px;
`;

const Link = styled.a`
  text-decoration: none;
  display: block;
`;

const StyledGridListTileBar = styled(GridListTileBar)`
  max-width: 300px;
`;

class PlaylistGrid extends Component {
  constructor(props) {
    super(props);
    this.state = {
      columns: 0,
    };
  }

  componentDidMount() {
    this.evaluateColumnsNumber();
    window.addEventListener('resize', this.evaluateColumnsNumber);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.evaluateColumnsNumber);
  }

  evaluateColumnsNumber = () => {
    this.setState({
      columns: Math.floor(window.innerWidth / 300) || 1,
    });
  }

  render() {
    const { playlists } = this.props;
    const { columns } = this.state;
    const cols = columns <= playlists.length ? columns : playlists.length;

    return (
      <Container>
        <GridList cellHeight={300} cols={cols}>
          {playlists.map(playlist => (
            <GridListTile key={playlist.images[0].url} cols={1}>
              <Link href={playlist.external_urls.spotify} target="_blank">
                <img src={playlist.images[0].url} alt="playlist" />
                <StyledGridListTileBar
                  title={playlist.name}
                  subtitle={playlist.owner.display_name}
                />
              </Link>
            </GridListTile>
          ))}
        </GridList>
      </Container>
    );
  }
}

PlaylistGrid.propTypes = {
  playlists: PropTypes.array,
};

PlaylistGrid.defaultProps = {
  playlists: [],
};

export default PlaylistGrid;
