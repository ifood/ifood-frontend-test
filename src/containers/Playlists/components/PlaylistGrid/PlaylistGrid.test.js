import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import PlaylistGrid from './PlaylistGrid';

describe('<PlaylistGrid />', () => {
  const mocks = {};

  beforeEach(() => {
    mocks.props = {
      playlists: [
        {
          name: 'playlist name1',
          owner: {
            display_name: 'some name 1',
          },
          images: [
            {
              url: 'some.url',
              width: 300,
              height: 300,
            },
          ],
          external_urls: {
            spotify: 'spotify url',
          },
        },
        {
          name: 'playlist name2',
          owner: {
            display_name: 'some name 2',
          },
          images: [
            {
              url: 'some.url2',
              width: 300,
              height: 300,
            },
          ],
          external_urls: {
            spotify: 'spotify url2',
          },
        },
      ],
    };
  });

  describe('render()', () => {
    it('renders the content correctly', () => {
      const wrapper = shallow(<PlaylistGrid {...mocks.props} />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('componentDidMount()', () => {
    it('should call evaluateColumnsNumber with expected parameters', () => {
      const wrapper = shallow(<PlaylistGrid {...mocks.props} />);
      wrapper.instance().evaluateColumnsNumber = jest.fn();
      wrapper.instance().componentDidMount();
      expect(wrapper.instance().evaluateColumnsNumber).toHaveBeenCalledWith();
    });

    it('should call window.addEventListener with expected parameters', () => {
      const wrapper = shallow(<PlaylistGrid {...mocks.props} />);
      const spy = jest.spyOn(window, 'addEventListener');
      wrapper.instance().componentDidMount();
      expect(spy).toHaveBeenCalledWith('resize', wrapper.instance().evaluateColumnsNumber);
    });
  });

  describe('componentWillUnmount()', () => {
    it('should call window.removeEventListener with expected parameters', () => {
      const wrapper = shallow(<PlaylistGrid {...mocks.props} />);
      const spy = jest.spyOn(window, 'removeEventListener');
      wrapper.instance().componentWillUnmount();
      expect(spy).toHaveBeenCalledWith('resize', wrapper.instance().evaluateColumnsNumber);
    });
  });

  describe('evaluateColumnsNumber()', () => {
    it('should call setState with expected params when width > 300', () => {
      const wrapper = shallow(<PlaylistGrid {...mocks.props} />);
      wrapper.instance().setState = jest.fn();
      window.innerWidth = 600;
      wrapper.instance().evaluateColumnsNumber();
      expect(wrapper.instance().setState).toHaveBeenCalledWith({ columns: 2 });
    });

    it('should call setState with expected params when width < 300', () => {
      const wrapper = shallow(<PlaylistGrid {...mocks.props} />);
      wrapper.instance().setState = jest.fn();
      window.innerWidth = 250;
      wrapper.instance().evaluateColumnsNumber();
      expect(wrapper.instance().setState).toHaveBeenCalledWith({ columns: 1 });
    });
  });
});
