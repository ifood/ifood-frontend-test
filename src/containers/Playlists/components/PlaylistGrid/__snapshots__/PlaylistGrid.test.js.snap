// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlaylistGrid /> render() renders the content correctly 1`] = `
<styled.div>
  <WithStyles(GridList)
    cellHeight={300}
    cols={2}
  >
    <WithStyles(GridListTile)
      cols={1}
      key="some.url"
    >
      <styled.a
        href="spotify url"
        target="_blank"
      >
        <img
          alt="playlist"
          src="some.url"
        />
        <Styled(WithStyles(GridListTileBar))
          subtitle="some name 1"
          title="playlist name1"
        />
      </styled.a>
    </WithStyles(GridListTile)>
    <WithStyles(GridListTile)
      cols={1}
      key="some.url2"
    >
      <styled.a
        href="spotify url2"
        target="_blank"
      >
        <img
          alt="playlist"
          src="some.url2"
        />
        <Styled(WithStyles(GridListTileBar))
          subtitle="some name 2"
          title="playlist name2"
        />
      </styled.a>
    </WithStyles(GridListTile)>
  </WithStyles(GridList)>
</styled.div>
`;
